### ISSUE TITLE
Conversion from valarray<double> to json fails to build

### ISSUE THREAD

**Issue by popizdeh** [nlohmann/json#702]
There seems to be a regression from 2.0.8 and 2.1.1. This no longer compiles.

```cpp
std::valarray<double> v;
nlohmann::json j;
j["test"] = v;
```

P.S. Is there any chance we can get a release with std::vector<bool> fix released?

**Comment by theodelrieu** [comment]
Oops, this is due to the `has_iterator` traits that check for a `iterator` type alias, which `valarray` doesn't provide. Sorry about that.

```cpp
namespace nlohmann
{
template <typename T>
struct adl_serializer<std::valarray<T>>
{
  static void to_json(json &j, std::valarray<T> const &v)
  {
    detail::external_constructor<detail::value_t::array>::construct(j, v);
  }

  static void from_json(json const &j, std::valarray<T> &v)
  {
    // checks omitted, not the best implem, never played with valarray myself
    auto arr = j.get<std::vector<T>>();
    v = std::valarray<T>(arr.data(), arr.size());
  }
};
}
```

As for the bug fix release, I don't know what @nlohmann thinks about it.

**Comment by nlohmann** [comment]
We need to fix this and add `std::valarray` support for the 3.x.x release. I did not plan to create a bug fix release for this.

**Comment by nlohmann** [comment]
Fixed with 9fd031b.
